File: leetcode/2_AddTwoNumbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

# The way to evaluate carry is elegant. A//=B returns an int instead of float. A //= 10 finds the carry bit
# Dont forget to count carry bit as condition of loop
# dont forget to update linked list

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        dummy = cur = ListNode(0)
        # carry bit also needs to count
        while l1 or l2 or carry:

            l1v = l1.val if l1 else 0
            l2v = l2.val if l2 else 0
            carry = l1v + l2v + carry
            cur.next = ListNode((carry) % 10)
            # dont forget this line to complete the loop
            cur = cur.next
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next

            carry //= 10
        return dummy.next

File: leetcode_java/2_AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode prev = new ListNode(0);
        ListNode head = prev;
        int carry = 0;
        while (l1 != null || l2 != null || carry != 0) {
            
            int l1v = 0;
            if (l1 != null) {
                l1v = l1.val;
                l1 = l1.next;
            }
            int l2v = 0;
            if (l2 != null) {
                l2v = l2.val;
                l2 = l2.next;
            }
            int sum = l1v + l2v + carry;
            carry = 0;
            if (sum > 9) {
                sum -= 10;
                carry = 1;
            }
            head.next = new ListNode(sum);        
            head = head.next;
        }
        
        return prev.next;
    }
}